p2p/transport/webrtc/datachannel.go

```diff
@@ -0,0 +1,341 @@
+	"github.com/pion/datachannel"
+	"github.com/pion/webrtc/v3"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+)
+
+var _ network.MuxedStream = &dataChannel{}
+
+const (
+	// maxMessageSize is limited to 16384 bytes in the SDP.
+	maxMessageSize uint64 = 16384
+	// Max message size limit in the SDP is limited to 16384 bytes.
+	// We keep a maximum of 2 messages in the buffer
+	maxBufferedAmount uint64 = 2 * maxMessageSize
+	// bufferedAmountLowThreshold and maxBufferedAmount are bound
+	// to a stream but congestion control is done on the whole
+	// SCTP association. This means that a single stream can monopolize
+	// the complete congestion control window (cwnd) if it does not
+	// read stream data and it's remote continues to send. We can
+	// add messages to the send buffer once there is space for 1 full
+	// sized message.
+	bufferedAmountLowThreshold uint64 = 16384
+
+	protoOverhead  int = 5
+	varintOverhead int = 2
+)
+
+const (
+	stateOpen uint32 = iota
+	stateReadClosed
+	stateWriteClosed
+	stateClosed
+)
+
+// Package pion detached data channel into a net.Conn
+// and then a network.MuxedStream
+type dataChannel struct {
+	channel       *webrtc.DataChannel
+	rwc           datachannel.ReadWriteCloser
+	laddr         net.Addr
+	raddr         net.Addr
+	readDeadline  *deadline
+	writeDeadline *deadline
+
+	closeWriteOnce sync.Once
+	closeReadOnce  sync.Once
+	resetOnce      sync.Once
+
+	state uint32
+
+	ctx            context.Context
+	cancel         context.CancelFunc
+	m              sync.Mutex
+	readBuf        []byte
+	writeAvailable chan struct{}
+	reader         protoio.Reader
+	writer         protoio.Writer
+}
+
+func newDataChannel(
+	channel *webrtc.DataChannel,
+	rwc datachannel.ReadWriteCloser,
+	pc *webrtc.PeerConnection,
+	laddr, raddr net.Addr) *dataChannel {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	result := &dataChannel{
+		channel:        channel,
+		rwc:            rwc,
+		laddr:          laddr,
+		raddr:          raddr,
+		readDeadline:   newDeadline(),
+		writeDeadline:  newDeadline(),
+		ctx:            ctx,
+		cancel:         cancel,
+		writeAvailable: make(chan struct{}),
+		reader:         protoio.NewDelimitedReader(rwc, 16384),
+		writer:         protoio.NewDelimitedWriter(rwc),
+		readBuf:        []byte{},
+	}
+
+	channel.SetBufferedAmountLowThreshold(bufferedAmountLowThreshold)
+	channel.OnBufferedAmountLow(func() {
+		result.writeAvailable <- struct{}{}
+	})
+
+	return result
+}
+
+func (d *dataChannel) processControlMessage(msg pb.Message) {
+	d.m.Lock()
+	defer d.m.Unlock()
+	if d.state == stateClosed {
+		return
+	}
+	if msg.Flag == nil {
+		return
+	}
+	switch msg.GetFlag() {
+	case pb.Message_FIN:
+		if d.state == stateWriteClosed {
+			d.Close()
+			return
+		}
+		d.state = stateReadClosed
+	case pb.Message_STOP_SENDING:
+		if d.state == stateReadClosed {
+			d.Close()
+			return
+		}
+		d.state = stateWriteClosed
+	case pb.Message_RESET:
+		d.channel.Close()
```

Comment:
The comment above might apply to this one instead.

Reply:
@marten-seemann I closed the comment because of the  following https://github.com/libp2p/specs/pull/412/files#r991894564. It is the same state machine implemented in Rust.